module/epochs/epoch_lookup.go

```diff
@@ -252,34 +258,67 @@ func (lookup *EpochLookup) EpochForViewWithFallback(view uint64) (uint64, error)
-// epoch fallback mode being triggered, we inject a fallback epoch.
+// the leader selection and cache static info for the epoch.
 func (lookup *EpochLookup) handleProtocolEvents(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
 	ready()
 
 	for {
 		select {
 		case <-ctx.Done():
 			return
-		case block := <-lookup.committedEpochsCh:
-			epoch := lookup.state.AtBlockID(block.ID()).Epochs().Next()
-			err := lookup.cacheEpoch(epoch)
+		case processEvtFn := <-lookup.epochEvents:
+			err := processEvtFn()
 			if err != nil {
 				ctx.Throw(err)
 			}
 		}
 	}
 }
 
-// EpochCommittedPhaseStarted informs the `committee.Consensus` that the block starting the Epoch Committed Phase has been finalized.
+// EpochExtended listens to `EpochExtended` protocol notifications. The notification is queued
+// for async processing by the worker. We must process _all_ `EpochExtended` notifications.
+func (lookup *EpochLookup) EpochExtended(epochCounter uint64, _ *flow.Header, extension flow.EpochExtension) {
+	lookup.epochEvents <- func() error {
+		return lookup.processEpochExtended(epochCounter, extension)
+	}
+}
+
+// EpochCommittedPhaseStarted ingests the respective protocol notifications. The notification is
+// queued for async processing by the worker. We must process _all_ `EpochCommittedPhaseStarted` notifications.
 func (lookup *EpochLookup) EpochCommittedPhaseStarted(_ uint64, first *flow.Header) {
-	lookup.committedEpochsCh <- first
+	lookup.epochEvents <- func() error {
+		return lookup.processEpochCommittedPhaseStarted(first)
+	}
+}
+
+// processEpochExtended processes the EpochExtended notification, which the Protocol
+// State emits when we finalize the first block whose Protocol State further extends the current
+// epoch. The next epoch should not be committed so far, because epoch extension are only added
+// when there is no subsequent epoch that we could transition into but the current epoch is nearing
+// its end. Specifically, we update the final view of the latest epoch range with the final view of the
+// current epoch, which will now be updated because the epoch has extensions.
+func (lookup *EpochLookup) processEpochExtended(epochCounter uint64, extension flow.EpochExtension) error {
+	err := lookup.epochs.extendLatestEpoch(epochCounter, extension)
+	if err != nil {
+		return err
+	}
+
+	return nil
 }
 
-// EpochFallbackModeTriggered passes the protocol event to the worker thread.
-func (lookup *EpochLookup) EpochFallbackModeTriggered(uint64, *flow.Header) {
-	lookup.epochFallbackIsTriggered.Store(true)
+// processEpochCommittedPhaseStarted processes the EpochCommittedPhaseStarted notification, which
+// the consensus component emits when we finalize the first block of the Epoch Committed phase.
+// Specifically, we cache the next epoch in the EpochLookup.
+// No errors are expected during normal operation.
+func (lookup *EpochLookup) processEpochCommittedPhaseStarted(first *flow.Header) error {
```

Comment:
I would also inline this to avoid nesting complexity.

Reply:
https://github.com/onflow/flow-go/pull/6247/commits/93ff1187b80854d81b81c5cded8d55c6a3b3b08f